### PR TITLE
Add new versions of gmsh

### PR DESCRIPTION
**Pull request by lrtfm**
Add new version of gmsh to the packages, including
```
4.5.5
4.5.6
4.6.0
4.7.0
4.7.1
```

In addition, a patch file is added for version 4.7.1 to the build error when `CGNS` option `CGNS_SCOPE_ENUMS` is on.
See [gmsh issue 1108](https://gitlab.onelab.info/gmsh/gmsh/-/issues/1108) and  [commit b0b3ece](https://gitlab.onelab.info/gmsh/gmsh/-/commit/b0b3ece698718bf1a79f50778d24d51c207f8727)

### CODE REVIEW

File: var/spack/repos/builtin/packages/gmsh/package.py [hunk 1]
```diff
@@ -30,6 +35,8 @@ class Gmsh(CMakePackage):
     version('2.12.0', sha256='7fbd2ec8071e79725266e72744d21e902d4fe6fa9e7c52340ad5f4be5c159d09')
     version('develop', branch='master', git='https://gitlab.onelab.info/gmsh/gmsh.git')
 
+    patch('gmsh-4.7.1-cgns-type.patch', sha256='bee4401602d114cbd011336261cd90a69d8cd80f4472f59ce9dfc65e8850dddb', when='@4.7.1')
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/gmsh/package.py]: If you do:
```python
patch('https://gitlab.onelab.info/gmsh/gmsh/-/commit/b0b3ece698718bf1a79f50778d24d51c207f8727.patch',
      sha256='...')
```
then you don't need to store such a large patch in Spack.

Comment (lrtfm) [reply var/spack/repos/builtin/packages/gmsh/package.py]: Thanks for that advice. But the patch commit cannot be applied to v4.7.1 directly. 
Any better suggestions? Or I can just delete this patch wait for the new versions to support CGNS.

Comment (adamjstewart) [reply var/spack/repos/builtin/packages/gmsh/package.py]: Hmm, in that case maybe just delete the patch and wait for a new release...

File: var/spack/repos/builtin/packages/gmsh/gmsh-4.7.1-cgns-type.patch [hunk 2]
```diff
@@ -0,0 +1,1006 @@
+From edf973c92d574ac91cb31e6a933be38113b8f78d Mon Sep 17 00:00:00 2001
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/gmsh/gmsh-4.7.1-cgns-type.patch]: You can delete this file now.

Comment (lrtfm) [reply var/spack/repos/builtin/packages/gmsh/gmsh-4.7.1-cgns-type.patch]: Thanks for your reminder. The file has been deleted.

**Review by adamjstewart — APPROVED** [review]

**Review by alalazo — APPROVED** [review]

**Pull request merged**